File: sdks/python/nodeblack_sdk.py
"""
NodeBlack Python SDK
Universal File Converter API Client

Installation:
    pip install requests

Usage:
    from nodeblack_sdk import NodeBlackClient
    
    client = NodeBlackClient("your-api-key")
    result = client.convert_file("image.png", "jpg")
    client.download_file(result["task_id"], "converted.jpg")
"""

import requests
import time
from typing import Optional, Dict, Any
from pathlib import Path

class NodeBlackClient:
    def __init__(self, api_key: str, base_url: str = "https://nodeblack.onrender.com"):
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.session = requests.Session()
        self.session.headers.update({"X-API-Key": api_key})
    
    def convert_file(self, file_path: str, target_format: str, timeout: int = 60) -> Dict[str, Any]:
        """
        Convert a file to target format
        
        Args:
            file_path: Path to input file
            target_format: Target format (e.g., 'jpg', 'png', 'mp3')
            timeout: Max wait time in seconds
            
        Returns:
            Dict with task_id and status
        """
        file_path = Path(file_path)
        
        if not file_path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")
        
        with open(file_path, 'rb') as f:
            files = {'file': f}
            params = {'target_format': target_format}
            
            response = self.session.post(
                f"{self.base_url}/api/convert",
                files=files,
                params=params
            )
            response.raise_for_status()
            
        result = response.json()
        task_id = result['task_id']
        
        # Wait for completion
        return self._wait_for_completion(task_id, timeout)
    
    def _wait_for_completion(self, task_id: str, timeout: int) -> Dict[str, Any]:
        """Wait for conversion to complete"""
        start_time = time.time()
        
        while time.time() - start_time < timeout:
            status = self.get_status(task_id)
            
            if status['status'] == 'ready':
                return {'task_id': task_id, 'status': 'completed'}
            elif status['status'] in ['failed', 'expired']:
                raise Exception(f"Conversion failed: {status.get('message', 'Unknown error')}")
            
            time.sleep(1)
        
        raise TimeoutError(f"Conversion timeout after {timeout} seconds")
    
    def get_status(self, task_id: str) -> Dict[str, Any]:
        """Get conversion status"""
        response = self.session.get(f"{self.base_url}/api/status/{task_id}")
        response.raise_for_status()
        return response.json()
    
    def download_file(self, task_id: str, output_path: str) -> str:
        """Download converted file"""
        response = self.session.get(f"{self.base_url}/api/download/{task_id}")
        response.raise_for_status()
        
        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        
        with open(output_path, 'wb') as f:
            f.write(response.content)
        
        return str(output_path)
    
    def get_supported_formats(self) -> Dict[str, Any]:
        """Get all supported formats"""
        response = self.session.get(f"{self.base_url}/api/formats")
        response.raise_for_status()
        return response.json()
    
    def list_files(self) -> Dict[str, Any]:
        """List all converted files"""
        response = self.session.get(f"{self.base_url}/api/files")
        response.raise_for_status()
        return response.json()

# Example usage
if __name__ == "__main__":
    # Initialize client
    client = NodeBlackClient("your-api-key-here")
    
    try:
        # Convert image
        result = client.convert_file("input.png", "jpg")
        print(f"✅ Conversion completed: {result}")
        
        # Download result
        output_file = client.download_file(result["task_id"], "output.jpg")
        print(f"✅ Downloaded: {output_file}")
        
    except Exception as e:
        print(f"❌ Error: {e}")

File: sdks/javascript/nodeblack-sdk.js
/**
 * NodeBlack JavaScript SDK
 * Universal File Converter API Client
 * 
 * Installation:
 *   npm install axios form-data
 * 
 * Usage:
 *   const NodeBlackClient = require('./nodeblack-sdk');
 *   
 *   const client = new NodeBlackClient('your-api-key');
 *   const result = await client.convertFile('image.png', 'jpg');
 *   await client.downloadFile(result.task_id, 'converted.jpg');
 */

const axios = require('axios');
const FormData = require('form-data');
const fs = require('fs');
const path = require('path');

class NodeBlackClient {
    constructor(apiKey, baseUrl = 'https://nodeblack.onrender.com') {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.client = axios.create({
            baseURL: this.baseUrl,
            headers: {
                'X-API-Key': apiKey
            }
        });
    }

    /**
     * Convert a file to target format
     * @param {string} filePath - Path to input file
     * @param {string} targetFormat - Target format (e.g., 'jpg', 'png', 'mp3')
     * @param {number} timeout - Max wait time in seconds (default: 60)
     * @returns {Promise<Object>} Result with task_id and status
     */
    async convertFile(filePath, targetFormat, timeout = 60) {
        if (!fs.existsSync(filePath)) {
            throw new Error(`File not found: ${filePath}`);
        }

        const formData = new FormData();
        formData.append('file', fs.createReadStream(filePath));

        try {
            const response = await this.client.post('/api/convert', formData, {
                params: { target_format: targetFormat },
                headers: formData.getHeaders()
            });

            const { task_id } = response.data;
            return await this._waitForCompletion(task_id, timeout);

        } catch (error) {
            throw new Error(`Conversion failed: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Wait for conversion to complete
     * @private
     */
    async _waitForCompletion(taskId, timeout) {
        const startTime = Date.now();

        while (Date.now() - startTime < timeout * 1000) {
            const status = await this.getStatus(taskId);

            if (status.status === 'ready') {
                return { task_id: taskId, status: 'completed' };
            } else if (['failed', 'expired'].includes(status.status)) {
                throw new Error(`Conversion failed: ${status.message || 'Unknown error'}`);
            }

            await new Promise(resolve => setTimeout(resolve, 1000));
        }

        throw new Error(`Conversion timeout after ${timeout} seconds`);
    }

    /**
     * Get conversion status
     * @param {string} taskId - Task ID
     * @returns {Promise<Object>} Status information
     */
    async getStatus(taskId) {
        try {
            const response = await this.client.get(`/api/status/${taskId}`);
            return response.data;
        } catch (error) {
            throw new Error(`Status check failed: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Download converted file
     * @param {string} taskId - Task ID
     * @param {string} outputPath - Output file path
     * @returns {Promise<string>} Path to downloaded file
     */
    async downloadFile(taskId, outputPath) {
        try {
            const response = await this.client.get(`/api/download/${taskId}`, {
                responseType: 'stream'
            });

            // Ensure output directory exists
            const outputDir = path.dirname(outputPath);
            if (!fs.existsSync(outputDir)) {
                fs.mkdirSync(outputDir, { recursive: true });
            }

            // Write file
            const writer = fs.createWriteStream(outputPath);
            response.data.pipe(writer);

            return new Promise((resolve, reject) => {
                writer.on('finish', () => resolve(outputPath));
                writer.on('error', reject);
            });

        } catch (error) {
            throw new Error(`Download failed: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Get all supported formats
     * @returns {Promise<Object>} Supported formats
     */
    async getSupportedFormats() {
        try {
            const response = await this.client.get('/api/formats');
            return response.data;
        } catch (error) {
            throw new Error(`Failed to get formats: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * List all converted files
     * @returns {Promise<Object>} List of files
     */
    async listFiles() {
        try {
            const response = await this.client.get('/api/files');
            return response.data;
        } catch (error) {
            throw new Error(`Failed to list files: ${error.response?.data?.detail || error.message}`);
        }
    }
}

module.exports = NodeBlackClient;

// Example usage
if (require.main === module) {
    (async () => {
        const client = new NodeBlackClient('your-api-key-here');

        try {
            // Convert image
            console.log('🔄 Converting file...');
            const result = await client.convertFile('input.png', 'jpg');
            console.log('✅ Conversion completed:', result);

            // Download result
            console.log('⬇️ Downloading file...');
            const outputFile = await client.downloadFile(result.task_id, 'output.jpg');
            console.log('✅ Downloaded:', outputFile);

        } catch (error) {
            console.error('❌ Error:', error.message);
        }
    })();
}